tests/unit/discovery/windows/conftest.py

```diff
@@ -0,0 +1,169 @@
+        key_id = id(key)
+        try:
+            result = EnumKey(key, at)
+            return result
+        except Exception as exception:
+            result = exception
+            raise result
+        finally:
+            enum_collect[key_id].append(result)
+
+    value_collect = defaultdict(dict)
+
+    def _v(key, value_name):
+        result = None
+        key_id = id(key)
+        try:
+            result = QueryValueEx(key, value_name)
+            return result
+        except Exception as exception:
+            result = exception
+            raise result
+        finally:
+            value_collect[key_id][value_name] = result
+
+    mocker.patch.object(winreg, "EnumKey", side_effect=_e)
+    mocker.patch.object(winreg, "QueryValueEx", side_effect=_v)
+    mocker.patch.object(winreg, "OpenKeyEx", side_effect=_c)
+
+    yield
+
+    print("")
+    print(f"hive_open = {hive_open}")
+    print(f"key_open = {dict(key_open.items())}")
+    print(f"value_collect = {dict(value_collect.items())}")
+    print(f"enum_collect = {dict(enum_collect.items())}")
+
+
+def _mock_pyinfo(major, minor, arch, exe):
+    """Return PythonInfo objects with essential metadata set for the given args"""
+    from virtualenv.discovery.py_info import PythonInfo, VersionInfo
+
+    info = PythonInfo()
+    info.base_prefix = str(Path(exe).parent)
+    info.executable = info.original_executable = info.system_executable = exe
+    info.implementation = "CPython"
+    info.architecture = arch
+    info.version_info = VersionInfo(major, minor, 0, "final", 0)
+    return info
+
+
+@pytest.fixture()
+def _populate_pyinfo_cache(monkeypatch):
+    """Add metadata to virtualenv.discovery.cached_py_info._CACHE for all (mocked) registry entries"""
+    import virtualenv.discovery.cached_py_info
+
+    # Data matches _mock_registry fixture
+    interpreters = [
+        ("ContinuumAnalytics", 3, 7, 32, "C:\\Users\\user\\Miniconda3\\python.exe", None),
+        ("ContinuumAnalytics", 3, 7, 64, "C:\\Users\\user\\Miniconda3-64\\python.exe", None),
+        ("PythonCore", 3, 6, 64, "C:\\Users\\user\\AppData\\Local\\Programs\\Python\\Python36\\python.exe", None),
+        ("PythonCore", 3, 6, 64, "C:\\Users\\user\\AppData\\Local\\Programs\\Python\\Python36\\python.exe", None),
+        ("PythonCore", 3, 5, 64, "C:\\Users\\user\\AppData\\Local\\Programs\\Python\\Python35\\python.exe", None),
+        ("PythonCore", 3, 6, 64, "C:\\Users\\user\\AppData\\Local\\Programs\\Python\\Python36\\python.exe", None),
+        ("PythonCore", 3, 7, 32, "C:\\Users\\user\\AppData\\Local\\Programs\\Python\\Python37-32\\python.exe", None),
+        ("PythonCore", 3, 9, 64, "C:\\Users\\user\\AppData\\Local\\Programs\\Python\\Python36\\python.exe", None),
+        ("PythonCore", 2, 7, 64, "C:\\Python27\\python.exe", None),
+        ("PythonCore", 3, 4, 64, "C:\\Python34\\python.exe", None),
+        ("CompanyA", 3, 3, 64, "Z:\\CompanyA\\Python\\3.3\\python.exe", None),
```

Comment:
Can we bump here the version numbers to more recent releases while we're here.

Reply:
Sure. It will obviously require quite a few changes in the test asserts which makes me a little nervous given we're also making changes to the code subject to test. 

Let me do it as one of the last commits.

Reply:
Done

Reply:
Done